modules/broker-amqp/src/main/java/org/wso2/broker/amqp/AmqpDeliverMessage.java

```diff
@@ -45,16 +51,31 @@
     private final AmqpChannel channel;
     private final Message message;
     private final String queueName;
+    private final Broker broker;
 
-    public AmqpDeliverMessage(Message message, ShortString consumerTag, AmqpChannel channel, String queueName) {
+    public AmqpDeliverMessage(Message message,
+                              ShortString consumerTag,
+                              AmqpChannel channel,
+                              String queueName,
+                              Broker broker) {
         this.message = message;
         this.consumerTag = consumerTag;
         this.channel = channel;
         this.queueName = queueName;
+        this.broker = broker;
     }
 
     public void write(ChannelHandlerContext ctx) {
-        if (!channel.isFlowEnabled()) {
+        if (channel.isClosed()) {
+            try {
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug("Requeueing message since subscriber is already closed. {}", message);
+                }
+                broker.requeue(queueName, message);
+            } catch (BrokerException e) {
+                LOGGER.error("Error while requeueing message {} for queue {}", message, queueName, e);
```

Comment:
What are the possible reasons to fail re-queue?

Reply:
In [Queue interface](https://github.com/wso2/message-broker/blob/master/modules/broker-core/src/main/java/org/wso2/broker/core/queue/Queue.java#L90) we are throwing a checked exception when enqueueing. Internally for requeueing we use an in-memory queue to keep requeued message. But with current implementation only error scenario is related to inserting data to DB which does not apply to in-memory queue. Therefore for this scenario, we should not get a checked exception even though interface has it. I didn't ignore the error since we could change the in-memory queue implementation later.